tests/test_configuration.py

```diff
@@ -81,3 +83,85 @@ def test_config_better_addresses_pushsafer():
     assert apprise_service_targets["nagios"]["device"] == "52|65|78"
     assert apprise_service_targets["nagios"]["priority"] == 2
     assert apprise_service_targets["tracking"]["device"] == "gs23"
+
+
+@patch.dict(os.environ, {"HOSTNAME": "example.com", "PORT": "3000", "USERNAME": "bob", "LOG_FILE": "/tmp/out.log"})
+def test_config_expand_variables():
+    """
+    Verify reading configuration file expands variables.
+    """
+    config = load_configuration(configfile_with_variables)
+    assert config.hostname == "example.com"
+    assert config.port == 3000
+    assert config.username == "bob"
+    assert config.password == "secret-password"
+    assert config.getdict("config:file", "targets")["mylog"][0] == "/tmp/out.log"
+
+
+@pytest.mark.parametrize(
+    "input, expected",
+    [
+        ("my-password", "my-password"),
+        ("$SRC_1:PASSWORD_1", "my-password"),
+        ("$SRC_1:PASSWORD_2", "super-secret"),
+        ("-->$SRC_1:PASSWORD_1<--", "-->my-password<--"),
+        ("$SRC_2:PASSWORD_1", "p4ssw0rd"),
+        ("$SRC_1:PÄSSWÖRD_3", "non-ascii-secret"),
+        ("${SRC_1:PASSWORD_1}", "my-password"),
+        ("${SRC_1:/path/to/password.txt}", "file-contents"),
+        ("${SRC_1:PASSWORD_1} ${SRC_1:PASSWORD_2}", "my-password super-secret"),
+        ("$SRC_1:PASSWORD_1 ${SRC_1:/path/to/password.txt} $SRC_1:PASSWORD_1", "my-password file-contents my-password"),
+        (
+            "${SRC_1:/path/to/password.txt} $SRC_1:PASSWORD_1 ${SRC_1:/path/to/password.txt}",
+            "file-contents my-password file-contents",
+        ),
+        ("/$SRC_1:PASSWORD_1/$SRC_1:PASSWORD_2/foo.txt", "/my-password/super-secret/foo.txt"),
+    ],
+)
+def test_expand_vars_ok(input, expected):
+    """
+    Verify that `expand_vars` expands variables in configuration.
+    """
+
+    def create_source(variables):
+        return lambda name: variables[name]
+
+    sources = {
+        "SRC_1": create_source(
+            {
+                "PASSWORD_1": "my-password",
+                "PASSWORD_2": "super-secret",
+                "PÄSSWÖRD_3": "non-ascii-secret",
+                "/path/to/password.txt": "file-contents",
+            }
+        ),
+        "SRC_2": create_source(
+            {
+                "PASSWORD_1": "p4ssw0rd",
+            }
+        ),
+    }
```

Comment:
`SRC_1` and `SRC_2` are two synthetic variable interpolation types like `ENV` or `FILE`, but only used within the test suite?

Reply:
Yes. I'm testing `expand_vars()` function here directly, bypassing concrete implementation of `ENV` and `FILE` sources, so I've replaced them with these fakes.